scripts/anvil_tools/batch_upsert_entities_standard.py

```diff
@@ -0,0 +1,148 @@
+
+    save_name = f"{filename_prefix}_batch_upsert_request.json"
+    with open(save_name, "w") as f:
+        f.write(request)
+
+
+def convert_string_to_list(input_string):
+    """Convert a given string into an array compatible with data model tsvs for array attributes."""
+
+    # remove single & double quotes, remove spaces, remove [ ], separate remaining string on commas (resulting in a list)
+    output_list = str(input_string).replace("'", '').replace('"', '').replace(" ", "").strip('[]').split(",")
+
+    return output_list
+
+
+def create_upsert_request(tsv, array_attr_cols=None):
+    """Generate the request body for batchUpsert API."""
+
+    # check tsv format: data model load tsv requirement "entity:table_name_id" or "membership:table_name_id" -> else exit
+    entity_type_col_name = tsv.columns[0]                               # entity:entity_name_id
+    entity_type = entity_type_col_name.rsplit("_", 1)[0].split(":")[1]  # entity_name
+
+    if not entity_type_col_name.startswith(("entity:", "membership:")):
+        print("Invalid tsv. The .tsv does not start with column entity:[table_name]_id or membership:[table_name]_id. Please correct and try again.")
+        return
+
+    # replace the "entity:col_name_id" with just "col_name" in df
+    # if not replaced, "attributeName" in the template_make_single_attr becomes "entity:entity_name_id" instead of just entity_name
+    # when the API request is made, its read as multiple columns with the "entity" prefix which is illegal
+    # this is specific just to the first column where the format is required for terra load tsv files
+    tsv.rename(columns={entity_type_col_name: entity_type}, inplace=True)
+
+    # templates
+    # for request body skeleton
+    template_req_body = """{"name":"VAR_ENTITY_ID", "entityType":"VAR_ENTITY_TYPE", "operations":[OPERATIONS_LIST]}"""
+
+    # for operations - create/update: list attribute, add list member to attribute if list, single attribute and value
+    template_make_list_attr = '{"op":"CreateAttributeValueList","attributeName":"VAR_ATTRIBUTE_LIST_NAME"},'
+    template_add_list_member = '{"op":"AddListMember","attributeListName":"VAR_ATTRIBUTE_LIST_NAME", "newMember":"VAR_LIST_MEMBER"},'
+    template_make_single_attr = '{"op":"AddUpdateAttribute","attributeName":"VAR_ATTRIBUTE_NAME", "addUpdateAttribute":"VAR_ATTRIBUTE_MEMBER"},'
+
+    # initiate string to capture all operation requests
+    all_attributes_request = []
+
+    for index, row in tsv.iterrows():
+        single_attribute_request = ''''''
+        # get the entity_id (row id - must be unique)
+        entity_id = str(row[0])
+
+        # if array columns/attributes
+        if array_attr_cols:
+            # for each array column
+            for col in array_attr_cols:
+                # add column name to template for a single attribute
+                single_attribute_request += template_make_list_attr.replace("VAR_ATTRIBUTE_LIST_NAME", col)
```

Comment:
this is a creative way to do this - you don't need to change it, but a more functional programming way to do it would have been to have a function for each operation like
```
def get_make_list_attr_request(var_attribute_list_name):
    return '{"op":"CreateAttributeValueList","attributeName":"' + var_attribute_list_name + '"},'

def get_add_list_member_request(var_attribute_list_name, var_list_member):
    return '{"op":"AddListMember","attributeListName":"' + var_attribute_list_name + '", "newMember":"' + var_list_member + '"},'
```
etc and one for make_single_attr as well.
to me that would be easier to parse than the replace string method. just food for thought!

Reply:
great idea - it does make it ALOT clearer